docling/utils/document_enrichment.py

```
@@ -331,14 +431,266 @@ def combine_windowed_toc(self, window_texts: List[str], *, joiner: str = "\n") -
+            start = end + 1 - overlap
+        return chunks
+
+    def _shrink_outline(self, toc: str, max_tokens: int) -> str:
+        """누적 outline이 max_tokens를 넘으면 상위 레벨만 남겨 축약(계층 컨텍스트 유지)."""
+        if not toc or max_tokens <= 0:
+            return toc
+        if self._estimate_tokens(toc) <= max_tokens:
+            return toc
+
+        lines = toc.split("\n")
+        title_lines = [l for l in lines if l.strip().startswith("TITLE:")]
```

Ruff (0.15.17)

[error] Ambiguous variable name: `l`

(E741)